katsdpcontroller/product_config.py

```diff
@@ -31,63 +82,1196 @@ class _Sensor(ABC):
+            except Exception as exc:
+                raise ValueError(f'Invalid source {i}: {exc}') from exc
+            sources.append(source)
+        return Simulation(
+            start_time=config.get('start_time'),
+            clock_ratio=config.get('clock_ratio', 1.0),
+            sources=sources
+        )
+
+
+class Stream:
+    """Base class for all streams."""
+
+    stream_type: ClassVar[str]
+    _class_sensors: ClassVar[Sequence[_Sensor]] = []
+    _valid_src_types: ClassVar[_ValidTypes] = set()
+
+    def __init__(self, name: str, src_streams: Sequence['Stream']) -> None:
+        self.name = name
+        self.src_streams = list(src_streams)
+
+    @classmethod
+    @abstractmethod
+    def from_config(cls: Type[_S],
+                    options: Options,
+                    name: str,
+                    config: Mapping[str, Any],
+                    src_streams: Sequence['Stream'],
+                    sensors: Mapping[str, Any]) -> _S: ...   # pragma: nocover
+
+
+class CamHttpStream(Stream):
+    """A cam.http stream."""
+
+    stream_type: ClassVar[str] = 'cam.http'
+
+    def __init__(self, name: str, *, url: yarl.URL) -> None:
+        super().__init__(name, [])
+        self.url = url
+
+    @classmethod
+    def from_config(cls,
+                    options: Options,
+                    name: str,
+                    config: Mapping[str, Any],
+                    src_streams: Sequence['Stream'],
+                    sensors: Mapping[str, Any]) -> 'CamHttpStream':
+        assert not src_streams
+        assert not sensors
+        return cls(name, url=yarl.URL(config['url']))
+
+
+class CbfStream:
+    """Mix-in for real CBF streams.
+
+    It exists mostly for isinstance tests.
+    """
+
+    url: yarl.URL
+    instrument_dev_name: str
+
+
+class AntennaChannelisedVoltageStreamBase(Stream):
+    """Base for both simulated and real antenna-channelised-voltage streams."""
+
+    def __init__(self, name: str, src_streams: Sequence[Stream], *,
+                 antennas: Iterable[str],
+                 band: str,
+                 n_channels: int,
+                 bandwidth: float,
+                 adc_sample_rate: float,
+                 centre_frequency: float,
+                 n_samples_between_spectra: int) -> None:
+        super().__init__(name, src_streams)
+        self.antennas = list(antennas)
+        self.band = band
+        self.n_channels = n_channels
+        self.bandwidth = bandwidth
+        self.centre_frequency = centre_frequency
+        self.adc_sample_rate = adc_sample_rate
+        self.n_samples_between_spectra = n_samples_between_spectra
+
+
+class AntennaChannelisedVoltageStream(CbfStream, AntennaChannelisedVoltageStreamBase):
+    """Real antenna-channelised-voltage stream."""
+
+    stream_type: ClassVar[str] = 'cbf.antenna_channelised_voltage'
+    _class_sensors: ClassVar[Sequence[_Sensor]] = [
+        _CBFSensor('n_chans', int),
```

Comment:
It's not possible any more to be 100% consistent everywhere because the SDP-CAM ICD uses "output_channels" rather than "output_chans", but I've renamed all instances of n_channels and n_channels_per_x to n_chans and n_chans_per_x to be a little less inconsistent.